Hot restart services refactoring

mmedenjak opened the pull request:
This is the first step in a small cleanup of hot restart classes and methods. 
- Hot restart related methods were moved from the NodeExtension into the InternalHotRestartService interface. 
- Hot restart service implementations were renamed.
The next step would be to refactor the InternalHotRestartServiceImpl so that separate functionalities are in separate classes.
Once we start adding non-hot-backup related public methods, we can create a new public implementation class and just delegate to the Hot backup service for backup related methods. For now all public methods are hot-backup related hence no need for a separate public service.

EE PR : https://github.com/hazelcast/hazelcast-enterprise/pull/1209

devOpsHazelcast commented:
Test PASSed.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/hotrestart/InternalHotRestartService.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Internal service for interacting with hot restart related functionalities (e.g. force and partial start)
+ */
+public interface InternalHotRestartService {
```
Is this class exposed to the public? If not, would it not make more sense to move it to com.hazelcast.internal.hotrestart? This way the com.hazelcast module remains clean + we get a good separation between public and private

mdogan replied:
What about `HotRestartServiceIntegrationService`? Since this is not related to implementing Hot Restart functionality, just to interact with actual Hot Restart impl.

mmedenjak replied:
That's right. Thus we would end up moving the entire package to `internal`. Perhaps that's too big of a change.

metanet replied:
I think current package structure is fine for now. We can discuss it later on and prepare a separate pr for class moves if we decide that way

pveentjer commented on hazelcast/src/main/java/com/hazelcast/internal/ascii/rest/HttpPostCommandProcessor.java:
```diff
@@ -233,7 +235,8 @@ private void handlePartialStart(HttpPostCommand command) throws UnsupportedEncod
             } else {
-                boolean success = node.getNodeExtension().triggerPartialStart();
+                final InternalHotRestartService hotRestartService = node.getNodeExtension().getInternalHotRestartService();
+                boolean success = hotRestartService != null && hotRestartService.triggerPartialStart();
```
Would it not be an idea to always have a hotRestartService but in case of not being enabled, that a NullHotRestartService is used? This way you never need to do null checks.

http://refactoring.com/catalog/introduceNullObject.html

metanet replied:
I guess we don't have many of these checks before so it was ok. For now, I am ok with both approaches. We can introduce a dummy hot restart service.

pveentjer commented:
Looks good so far. Keep up the good work!

mdogan commented on hazelcast/src/main/java/com/hazelcast/instance/Node.java:
```diff
@@ -666,10 +669,13 @@ public SplitBrainJoinMessage createSplitBrainJoinMessage() {
         final Credentials credentials = (withCredentials && securityContext != null)
                 ? securityContext.getCredentialsFactory().newCredentials() : null;
+        final InternalHotRestartService hotRestartService = nodeExtension.getInternalHotRestartService();
+        final Set<String> excludedMemberUuids = hotRestartService != null
```
Instead of `null` checking everywhere, can we have a nop implementation of `InternalHotRestartService` in OSS? 
So `nodeExtension.getInternalHotRestartService()` never returns `null`...

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6319/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/6321/

devOpsHazelcast commented:
Test PASSed.

mmedenjak commented:
@pveentjer @mdogan @metanet I've :
- changed the class name `InternalHotRestartService` to `HotRestartIntegrationService`
- changed the class name `InternalHotBackupService` to `HotBackupService`
- added noop implementations of these classes

Most Hot restart functionality is in the `com.hazelcast.spi.hotrestart` package and most of it is private. Although the package name could be changed to internal we will postpone this for an another PR.

I think this addresses most of the comments. If noone has any objections today, I will merge.

devOpsHazelcast commented:
Test PASSed.
